src/net/socket.rs

```diff
@@ -0,0 +1,229 @@
+            }
+            if let Err(e) = self.process_events(&mut events) {
+                error!("Error processing events: {:?}", e);
+            }
+
+            for packet in packet_receiver.try_iter() {
+                // XXX: I'm fairly certain this isn't exactly safe. I'll likely need to add some
+                // handling for when the socket is blocked on send. Worth some more research.
+                // Alternatively, I'm sure the Tokio single threaded runtime does handle this for us
+                // so maybe it's work switching to that while providing the same interface?
+                if let Err(e) = self.send_to(packet) {
+                    error!("Error sending packet: {:?}", e);
+                }
+            }
+        }
+    }
+
+    /// Iterate through all of the idle connections based on `idle_connection_timeout` config and
+    /// remove them from the active connections. For each connection removed, we will send a
+    /// `SocketEvent::TimeOut` event to the `event_sender` channel.
+    fn handle_idle_clients(&mut self) {
+        let idle_addresses = self
+            .connections
+            .idle_connections(self.config.idle_connection_timeout);
+
+        for address in idle_addresses {
+            self.connections.remove_connection(&address);
+            if let Err(err) = self.event_sender.send(SocketEvent::TimeOut(address)) {
+                error!("Error sending timeout: {:?}", err);
+            }
+        }
+    }
+
+    // Process events received from the mio socket.
+    fn process_events(&mut self, events: &mut Events) -> NetworkResult<()> {
+        for event in events.iter() {
+            match event.token() {
+                SOCKET => {
+                    if event.readiness().is_readable() {
+                        loop {
+                            match self.recv_from() {
+                                Ok(Some(packet)) => {
+                                    if let Err(err) =
+                                        self.event_sender.send(SocketEvent::Packet(packet))
+                                    {
+                                        error!("Error sending packet to caller: {:?}", err);
+                                    }
+                                }
+                                Ok(None) => continue,
+                                Err(ref err) => match err.kind() {
+                                    NetworkErrorKind::IOError(io_err)
+                                        if io_err.kind() == io::ErrorKind::WouldBlock =>
+                                    {
+                                        break;
+                                    }
+                                    _ => error!("Error receiving from socket: {:?}", err),
+                                },
+                            };
+                        }
+                    }
+                }
+                _ => unreachable!(),
```

Comment:
I think we can add a string here to be more clear if this ever does hit for some odd reason. Better than just an unreachable panic.